src/issues/shareProviders.ts

```diff
@@ -0,0 +1,242 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { Commit, Remote, Repository } from '../api/api';
+import { GitApiImpl } from '../api/api1';
+import { fromReviewUri, Schemes } from '../common/uri';
+import { FolderRepositoryManager } from '../github/folderRepositoryManager';
+import { RepositoriesManager } from '../github/repositoriesManager';
+import { getRepositoryForFile } from '../github/utils';
+import { encodeURIComponentExceptSlashes, getBestPossibleUpstream, getOwnerAndRepo, getSimpleUpstream, getUpstreamOrigin, rangeString } from './util';
+
+export class ShareProviderManager implements vscode.Disposable {
+	private disposables: vscode.Disposable[] = [];
+
+	constructor(repositoryManager: RepositoriesManager, gitAPI: GitApiImpl) {
+		if (!vscode.window.registerShareProvider) {
+			return;
+		}
+
+		this.disposables.push(
+			new GitHubDevShareProvider(repositoryManager, gitAPI),
+			new GitHubPermalinkShareProvider(repositoryManager, gitAPI),
+			new GitHubHeadLinkShareProvider(repositoryManager, gitAPI)
+		);
+	}
+
+	dispose() {
+		this.disposables.forEach((d) => d.dispose());
+	}
+}
+
+const supportedSchemes = [Schemes.File, Schemes.Review, Schemes.Pr, Schemes.PRNode, Schemes.FileChange, Schemes.GithubPr, Schemes.VscodeVfs, Schemes.Comment];
```

Comment:
I don't think we need all these schemes. Specifically, I don't think we need the following:
- `Comment` is from the comment widget. 
- `PRNode` is for a tree item that isn't related to a file. 
- `GitHubPr` is for the diffs you can see while you're creating a PR.
- `FileChange` is only used in a tree item `context` value so I don't actually think we will ever encouter it.

Reply:
Thanks! I think we'd want to support sharing links from inside a diff so I will leave `Schemes.GithubPr` in. I would also like to support share in tree items via a new `shareableItem` optional property in tree items, but that depends on us having support for asynchronous context menu items from Electron, so I will take that out until we have a better story figured out there.

Reply:
PR diffs are `Schemes.Pr` and `Schemes.Review`. `Schemes.GitHubPr` is just the diff you can see when _creating_ a PR.